File: 219-contains-duplicate-ii/contains-duplicate-ii.cpp
class Solution {
public:
    bool containsNearbyDuplicate(vector<int>& nums, int k) {
        int i=0, j=0;
        unordered_set<int> check;
        while(j < nums.size()){
            int len = j-i;
            if (len <= k){
                if (check.count(nums[j])){
                    return true;
                }
                check.insert(nums[j]);
                j+=1;
            } else {
                check.erase(nums[i]);
                i++;
            }
        }
        return false;
    }
};

File: 219-contains-duplicate-ii/contains-duplicate-ii.go
func containsNearbyDuplicate(nums []int, k int) bool {
    if k == 0 {
        return false
    }
    j := 0
    seen := map[int]int{}
    for i := 0; i < len(nums); i++ {
        curr := nums[i]

        win_len := i - j

        if win_len > k {
            // now shrink the window 
            for win_len > k {
                temp := nums[j]
                delete(seen, temp)
                j++
                win_len = i - j
            }
        }

        // Now check if there are duplicates or not
        if _, ok := seen[curr]; ok {
            return true
        }
        seen[nums[i]]++
    }

    return false
}